File: 0/img_process.py
# Assignment 0
# Processing an image to produce an image with averaged rows

# Uses 'python 3.x'
# Dependencies 'sys' 'cv2' 'numpy' 'argparse'

# Rajbir Malik
# 2017CS10416

import numpy as np
import cv2 as cv
import argparse
import sys


# Function for command line arguments
def set_parser():
    # Creating a parser
    parser = argparse.ArgumentParser(
            description="Assignment-0 (Command Line Arguments)")
    # Adding -in
    parser.add_argument('-in', action="store", dest="_in", default=None, type=str, help='Input-file relative path.')
    # Adding -out
    parser.add_argument('-out', action="store", dest="_out", default=None, type=str, help='Output-file relative path.')
    return parser


# Processing Function
def process(img):
    img = np.array(img, dtype=np.uint64)
    # Slow, single line solution
    new_img = []
    row_len = img.shape[1]
    for row in img:
        avg = row.sum(axis=0)/row_len
        new_img.append([avg]*row_len)
    return np.array(new_img)


# Main
if __name__ == '__main__':

    # Setting up parser
    parser = set_parser()

    # Main program
    try:

        # Parsing
        parse = parser.parse_args(sys.argv[1:])

        # Reading
        if parse._in:
            img = cv.imread(parse._in)
        else:
            img = cv.imread(input('Enter relative file path : '))

        # Function (Processing)
        img = process(img)

        # Writing
        if parse._out:
            img = cv.imwrite(parse._out, img)
        else:
            print('Image written at relative path ./final.jpeg')
            cv.imwrite('./final.jpeg', img)

    # Catching exceptions (while reading, writing...)
    except Exception as _e:
        print(_e.__doc__())

File: 0/img_process.cpp
#include <opencv2/opencv.hpp>
#include <string>
#include <exception>

/*
! Program for reading an image, processing it and writing a new image formed by averaging the rows.

* Parameters
    * @param argv[1] -> Relative path of input image
    * @param argv[2] -> Relative path of output image
*/

// Main
int main(int argc, char** argv) {
    // Input and output name strings
    std::string _in, _out{"final.jpeg"} ;
    // Naming state
    switch(argc) {
        // No cli-args
        case 1: {
            std::cout << "Enter the input file name : ";
            std::cin >> _in;
            break;
            }
        // Input name args
        case 2: {
            _in = argv[1];
            break;
            }
        // Input and output name args
        case 3: {
            _in = argv[1];
            _out = argv[2];
            break;
            }
        //excess args
        default : {
            std::cout << "Only 2 arguments accepted. (_in & _out)\n";
            _in = argv[1];
            _out = argv[2];
            break;
            }
    }
    // Image variable
    cv::Mat img;
    try {
        // Loading the image
        img = cv::imread(_in,1);
        // Exiting if image doesn't exist
        if(img.empty())
            throw std::runtime_error("File Doesn't exist\nImage is empty\n");
    } 
    catch(const std::runtime_error& e) {
        // Image doesn't exist
        std::cerr << e.what();
    } 
    catch(...) {
        // Unknown error in reading (exit code 1)
        std::cerr << "Problem in reading image.\nExiting...\n";
        exit(1);
    }

    // New image matrix
    cv::Size_ size = img.size();
    cv::Mat new_img = cv::Mat::zeros(size, img.type());

    // Processing the image row wise
    for(size_t i {}; i < size.height; i++) {
        int bgr[] {0, 0, 0};
        // Summing the row
        for(size_t j {}; j < size.width; j++) {
            uchar* x = img.ptr<uchar>(i, j);
            bgr[0] += (int)x[0];
            bgr[1] += (int)x[1];
            bgr[2] += (int)x[2];
        }
        // Averaging values
        bgr[0] /= size.width;
        bgr[1] /= size.width;
        bgr[2] /= size.width;
        // Setting the row of new image
        for(size_t j {}; j < size.width; j++) {
            new_img.at<cv::Vec3b>(i, j) = cv::Vec3b((uchar)bgr[0], (uchar)bgr[1], (uchar)bgr[2]);
        }
    }
    // Writing the new created image
    cv::imwrite(_out, new_img);
    return 0;
}
